File: UISamples/Bit6UIFullDemo-ObjC/ChatDemo/ChatsTableViewController.m
//
//  ChatsViewController.m
//  Bit6ChatDemo
//
//  Created by Carlos Thurber Boaventura on 04/01/14.
//  Copyright (c) 2014 Bit6. All rights reserved.
//

#import "ChatsTableViewController.h"
#import "ConversationDetailsTableViewController.h"

@interface ChatsTableViewController () <BXULocationViewControllerDelegate, BXUImageViewControllerDelegate>

@end

@implementation ChatsTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIButton *button = [UIButton buttonWithType:UIButtonTypeInfoLight];
    [button addTarget:self action:@selector(showDetailInfo) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *detailButton = [[UIBarButtonItem alloc] initWithCustomView:button];
    
    self.navigationItem.rightBarButtonItems = @[detailButton, self.callButtonItem];
}

#pragma mark - BXUMessageTableViewDelegate

- (void)didSelectImageMessage:(nonnull Bit6Message*)message
{
    BXUImageViewController *imageVC = [BXUImageViewController new];
    imageVC.delegate = self;
    [imageVC setMessage:message];
    
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:imageVC];
    navController.navigationBar.barStyle = self.navigationController.navigationBar.barStyle;
    if (self.navigationController.navigationBar.barTintColor) {
        navController.navigationBar.barTintColor = self.navigationController.navigationBar.barTintColor;
    }
    if (self.navigationController.navigationBar.titleTextAttributes) {
        navController.navigationBar.titleTextAttributes = self.navigationController.navigationBar.titleTextAttributes;
    }
    if (self.navigationController.navigationBar.tintColor) {
        navController.navigationBar.tintColor = self.navigationController.navigationBar.tintColor;
    }
    
    navController.navigationBar.translucent = NO;
    navController.toolbar.translucent = NO;
    navController.modalPresentationStyle = UIModalPresentationFormSheet;
    [self presentViewController:navController animated:YES completion:nil];
}

- (void)didSelectVideoMessage:(nonnull Bit6Message*)message
{
    [Bit6 playVideoFromMessage:message viewController:self.navigationController];
}

- (void)didSelectLocationMessage:(nonnull Bit6Message*)message
{
    BXULocationViewController *locationVC = [BXULocationViewController new];
    locationVC.delegate = self;
    locationVC.coordinate = message.location;
    
    NSMutableArray *actions = [NSMutableArray arrayWithCapacity:2];
    if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"comgooglemaps-x-callback://"]]) {
        [actions addObject:@"Google Maps"];
    }
    if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"waze://"]]) {
        [actions addObject:@"Waze"];
    }
    locationVC.actions = actions;
    
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:locationVC];
    navController.navigationBar.barStyle = self.navigationController.navigationBar.barStyle;
    if (self.navigationController.navigationBar.barTintColor) {
        navController.navigationBar.barTintColor = self.navigationController.navigationBar.barTintColor;
    }
    if (self.navigationController.navigationBar.titleTextAttributes) {
        navController.navigationBar.titleTextAttributes = self.navigationController.navigationBar.titleTextAttributes;
    }
    if (self.navigationController.navigationBar.tintColor) {
        navController.navigationBar.tintColor = self.navigationController.navigationBar.tintColor;
    }
    
    navController.navigationBar.translucent = NO;
    navController.toolbar.translucent = NO;
    navController.modalPresentationStyle = UIModalPresentationFormSheet;
    [self presentViewController:navController animated:YES completion:nil];
}

#pragma mark - BXULocationViewControllerDelegate

- (void)dismissLocationViewController:(nonnull BXULocationViewController*)locationViewController
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)didSelectLocationAction:(nonnull NSString*)action coordinate:(CLLocationCoordinate2D)coordinate
{
    if ([action isEqualToString:@"Google Maps"]) {
        NSString *urlStr = [NSString stringWithFormat:@"comgooglemaps-x-callback://?q=%f,%f&zoom=14&x-success=bit6fulldemo://&x-source=Bit6Demo", coordinate.latitude, coordinate.longitude];
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlStr]];
    }
    else if ([action isEqualToString:@"Waze"]) {
        NSString *urlStr = [NSString stringWithFormat:@"waze://?ll=%f,%f&navigate=yes", coordinate.latitude, coordinate.longitude];
        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlStr]];
    }
}

#pragma mark - BXUImageViewControllerDelegate

- (void)dismissImageViewController:(BXUImageViewController *)imageViewController
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Navigation

- (void) showDetailInfo
{
    [self performSegueWithIdentifier:@"showDetails" sender:nil];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"showDetails"]) {
        ConversationDetailsTableViewController *obj = segue.destinationViewController;
        obj.conversation = [Bit6Conversation conversationWithAddress:self.address];
    }
}

@end

File: UISamples/Bit6UIFullDemo-Swift/Bit6ChatDemo-Swift/ChatsTableViewController.swift
//
//  ChatsTableViewController.swift
//  Bit6ChatDemo-Swift
//
//  Created by Carlos Thurber Boaventura on 07/08/14.
//  Copyright (c) 2014 Bit6. All rights reserved.
//

import UIKit
import MobileCoreServices
import Bit6
import Bit6UI

class ChatsTableViewController: BXUMessageTableViewController, BXULocationViewControllerDelegate, BXUImageViewControllerDelegate {

    override func viewDidLoad() {
        super.viewDidLoad()
    
        let button = UIButton(type: .InfoLight)
        button.addTarget(self, action:#selector(ChatsTableViewController.showDetailInfo), forControlEvents:.TouchUpInside)
        let detailButton = UIBarButtonItem(customView: button)

        self.navigationItem.rightBarButtonItems = [detailButton, self.callButtonItem]
    }
    
    // MARK: - BXUMessageTableViewDelegate
    
    override func didSelectImageMessage(message: Bit6Message) {
        let imageVC = BXUImageViewController()
        imageVC.delegate = self
        imageVC.setMessage(message)
        
        let navController = UINavigationController(rootViewController:imageVC)
        navController.navigationBar.barStyle = (self.navigationController?.navigationBar.barStyle)!

        if let barTintColor = self.navigationController?.navigationBar.barTintColor {
            navController.navigationBar.barTintColor = barTintColor
        }
        if let titleTextAttributes = self.navigationController?.navigationBar.titleTextAttributes {
            navController.navigationBar.titleTextAttributes = titleTextAttributes
        }
        if let tintColor = self.navigationController?.navigationBar.tintColor {
            navController.navigationBar.tintColor = tintColor
        }
        
        navController.navigationBar.translucent = false
        navController.toolbar.translucent = false
        navController.modalPresentationStyle = .FormSheet
        presentViewController(navController, animated:true, completion:nil)
        
    }
    
    override func didSelectVideoMessage(message: Bit6Message) {
        Bit6.playVideoFromMessage(message, viewController:self.navigationController!)
    }
    
    override func didSelectLocationMessage(message: Bit6Message) {
        let locationVC = BXULocationViewController()
        locationVC.delegate = self
        locationVC.coordinate = message.location;
        
        var actions = [String]()
        if UIApplication.sharedApplication().canOpenURL(NSURL(string:"comgooglemaps-x-callback://")!) {
            actions.append("Google Maps")
        }
        if UIApplication.sharedApplication().canOpenURL(NSURL(string:"waze://")!) {
            actions.append("Waze")
        }
        locationVC.actions = actions;
        
        let navController = UINavigationController(rootViewController:locationVC)
        navController.navigationBar.barStyle = (self.navigationController?.navigationBar.barStyle)!
        
        if let barTintColor = self.navigationController?.navigationBar.barTintColor {
            navController.navigationBar.barTintColor = barTintColor
        }
        if let titleTextAttributes = self.navigationController?.navigationBar.titleTextAttributes {
            navController.navigationBar.titleTextAttributes = titleTextAttributes
        }
        if let tintColor = self.navigationController?.navigationBar.tintColor {
            navController.navigationBar.tintColor = tintColor
        }
        
        navController.navigationBar.translucent = false
        navController.toolbar.translucent = false
        navController.modalPresentationStyle = .FormSheet
        presentViewController(navController, animated:true, completion:nil)
    }
    
    // MARK: - BXULocationViewControllerDelegate
    
    func dismissLocationViewController(locationViewController: BXULocationViewController) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func didSelectLocationAction(action: String, coordinate: CLLocationCoordinate2D) {
        if action == "Google Maps" {
            let urlStr = "comgooglemaps-x-callback://?q=\(coordinate.latitude),\(coordinate.longitude)&zoom=14&x-success=bit6fulldemo://&x-source=Bit6Demo"
            UIApplication.sharedApplication().openURL(NSURL(string: urlStr)!)
        }
        else if action == "Waze" {
            let urlStr = "waze://?ll=\(coordinate.latitude),\(coordinate.longitude)&navigate=yes"
            UIApplication.sharedApplication().openURL(NSURL(string: urlStr)!)
        }
    }
    
    // MARK: -BXUImageViewControllerDelegate
    
    func dismissImageViewController(imageViewController: BXUImageViewController) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    // MARK: - Navigation
    
    func showDetailInfo() {
        self.performSegueWithIdentifier("showDetails", sender:nil)
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject!) {
        if segue.identifier == "showDetails" {
            let obj = segue.destinationViewController as! ConversationDetailsTableViewController
            obj.conversation = Bit6Conversation(address:self.address)
        }
    }
}
